File: framework/languages/java/samples/java/GameQuest/Server/GameApi/src/main/java/systems/zlink/samples/gamequest/server/gameapi/handlers/RebuildProjectionHandler.java
package systems.zlink.samples.gamequest.server.gameapi.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.gamequest.server.gameapi.contracts.SelfCheckMessages;
import systems.zlink.samples.gamequest.server.gameapi.store.GameQuestStore;

/** Self-check: rebuilds a quest projection entry from the stored event stream. */
@ZLinkHandlerGroup("gameapi")
public final class RebuildProjectionHandler
    implements ZLinkRequestHandler<
        SelfCheckMessages.RebuildProjectionReq,
        SelfCheckMessages.RebuildProjectionRes> {
    private final GameQuestStore store;

    public RebuildProjectionHandler(GameQuestStore store) {
        this.store = store;
    }

    @Override
    public SelfCheckMessages.RebuildProjectionRes handle(
        SelfCheckMessages.RebuildProjectionReq request, ZLinkRequestContext context) {
        return new SelfCheckMessages.RebuildProjectionRes(
            store.rebuildProjection(request.playerId(), request.questId()));
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/GameApi/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/gameapi/handlers/RebuildProjectionHandler.kt
package systems.zlink.samples.kotlin.gamequest.server.gameapi.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.gamequest.server.gameapi.contracts.RebuildProjectionReq
import systems.zlink.samples.kotlin.gamequest.server.gameapi.contracts.RebuildProjectionRes
import systems.zlink.samples.kotlin.gamequest.server.gameapi.store.GameQuestStore

/** Self-check: rebuilds a quest projection entry from the stored event stream. */
@ZLinkHandlerGroup("gameapi")
class RebuildProjectionHandler(
    private val store: GameQuestStore,
) : ZLinkSuspendingRequestHandler<RebuildProjectionReq, RebuildProjectionRes> {
    override suspend fun handle(request: RebuildProjectionReq, context: ZLinkRequestContext): RebuildProjectionRes =
        RebuildProjectionRes(store.rebuildProjection(request.playerId, request.questId))
}
